File: 0735 asteroid-collision/solution.java
class Solution {
    public int[] asteroidCollision(int[] as) {
        int[] st = new int[as.length];
        int top = -1;
        for (int i : as) {
            boolean a = true;
            while (a && top >= 0 && st[top] > 0 && i < 0) {
                if (st[top] < -i)
                    top--;
                else if (st[top] == -i) {
                    top--;
                    a = false;
                } else {
                    a = false;
                }
            }
            if (a)
                st[++top] = i;
        }

        int[] ans = new int[top + 1];
        for (int i = 0; i <= top; i++) {
            ans[i] = st[i];
        }
        return ans;
    }
}

File: 0735 asteroid-collision/solution.rs
impl Solution {
    pub fn asteroid_collision(ar: Vec<i32>) -> Vec<i32> {
        let mut st: Vec<i32> = Vec::new();
        for &val in ar.iter() {
            let mut a = true;
            while !st.is_empty() && *st.last().unwrap() > 0 && val < 0 {
                let top = *st.last().unwrap();
                if top < -val {
                    st.pop();
                    continue;
                } else if top == -val {
                    st.pop();
                    a=false;
                } else {
                    a=false;
                }
                break;
            }
            if a {
                st.push(val);
            }
        }
        st
    }
}
